// loon/src/mvc/MiddlewareMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MiddlewareMetadata {
    get type() {
        return this._type;
    }
    get order() {
        return this._order;
    }
    get baseUrl() {
        return this._baseUrl;
    }
    get isErrorMiddleware() {
        return this._isErrorMiddleware;
    }
    set isErrorMiddleware(value) {
        this._isErrorMiddleware = value;
    }
    get handler() {
        return this._handler;
    }
    set handler(handler) {
        if (this._isErrorMiddleware) {
            handler.isErrorHandler = true;
        }
        this._handler = handler;
    }
    constructor(type, options) {
        this._type = type;
        this.init(options);
    }
    init(options) {
        this._order = Number.MAX_SAFE_INTEGER;
        this._baseUrl = "/";
        this._isErrorMiddleware = false;
        if (options && typeof options.order !== 'undefined') {
            this._order = options.order;
        }
        if (options && typeof options.baseUrl !== 'undefined') {
            this._baseUrl = options.baseUrl;
        }
        return this;
    }
}
exports.MiddlewareMetadata = MiddlewareMetadata;
