// loon/lib/mvc/HandlerMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HandlerMetadata {
    get params() {
        return this._params;
    }
    get isErrorHandler() {
        return this._isErrorHandler;
    }
    get type() {
        return this._type;
    }
    get actionName() {
        return this._actionName;
    }
    get httpMethodAndPaths() {
        return this._httpMethodAndPaths;
    }
    set params(value) {
        this._params = value;
    }
    set isErrorHandler(value) {
        this._isErrorHandler = value;
    }
    constructor(type, actionName, params, isErrorHandler) {
        this._type = type;
        this._actionName = actionName;
        if (typeof params !== 'undefined') {
            this._params = params;
        }
        else {
            this._params = new Map();
        }
        if (typeof isErrorHandler !== 'undefined') {
            this._isErrorHandler = isErrorHandler;
        }
        else {
            this._isErrorHandler = false;
        }
        this._httpMethodAndPaths = [];
    }
}
exports.HandlerMetadata = HandlerMetadata;
