// loon/lib/mvc/HandlerParamMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HandlerParamMetadata {
    get required() {
        return this._required;
    }
    get type() {
        return this._type;
    }
    get paramType() {
        return this._paramType;
    }
    get index() {
        return this._index;
    }
    get expression() {
        return this._expression;
    }
    get actionName() {
        return this._actionName;
    }
    get returnType() {
        return this._returnType;
    }
    constructor(type, actionName, index, returnType, paramType, expression) {
        this._type = type;
        this._actionName = actionName;
        this._returnType = returnType;
        this._index = index;
        if (typeof paramType !== 'undefined') {
            this._paramType = paramType;
        }
        if (typeof expression !== 'undefined') {
            this._expression = expression;
        }
    }
    setOptions(options) {
        if (options && typeof options.required !== 'undefined') {
            this._required = options.required;
        }
        else {
            this._required = false;
        }
        if (options && typeof options.defaultValue !== 'undefined') {
            this._defaultValue = options.defaultValue;
        }
    }
}
exports.HandlerParamMetadata = HandlerParamMetadata;
