// loon/lib/converter/PropertyMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PropertyMetadata {
    get type() {
        return this._type;
    }
    get objectProperty() {
        return this._objectProperty;
    }
    get klassProperty() {
        return this._klassProperty;
    }
    get propertyType() {
        return this._propertyType;
    }
    get converter() {
        return this._converter;
    }
    set converter(value) {
        this._converter = value;
    }
    get serialize() {
        return this._serialize;
    }
    get deserialize() {
        return this._deserialize;
    }
    get baseType() {
        return this._baseType;
    }
    set baseType(value) {
        this._baseType = value;
    }
    constructor(type, klassProperty, objectProperty, propertyType, serialize, deserialize) {
        this._type = type;
        this._objectProperty = objectProperty;
        this._klassProperty = klassProperty;
        this._propertyType = propertyType;
        this._serialize = serialize;
        this._deserialize = deserialize;
    }
}
exports.PropertyMetadata = PropertyMetadata;
